engine/src/main/scala/cromwell/engine/workflow/tokens/TokenQueue.scala

```diff
@@ -31,23 +34,41 @@ final case class TokenQueue(queue: Queue[ActorRef], private [tokens] val pool: T
     *
     * @return the new token queue
     */
-  def enqueue(actor: ActorRef): TokenQueue = {
-    copy(queue = queue.enqueue(actor))
+  def enqueue(placeholder: TokenQueuePlaceholder): TokenQueue = {
+    copy(queue = queue.enqueue(placeholder))
   }
 
   /**
-    * Returns a DequeuedActor if there's an element available, None otherwise.
+    * Returns a dequeue'd actor if one exists and there's a token available for it
+    * Returns an updated token queue based on this request (successful queuees get removed, hogs get sent to the back)
     */
-  def dequeueOption: Option[DequeuedActor] = {
-    for {
-      actorAndNewQueue <- queue.dequeueOption
-      (actor, newQueue) = actorAndNewQueue
-      lease <- pool.tryAcquire()
-    } yield DequeuedActor(LeasedActor(actor, lease), copy(queue = newQueue))
+  def dequeue: DequeueResult = {
+    queue.dequeueOption match {
+      case Some(actorAndNewQueue) =>
+        val (placeholder, newQueue) = actorAndNewQueue
+        val poolAcquisitionResult = pool.tryAcquire(placeholder.hogGroup)
+        shuffleQueueAndGetLease(poolAcquisitionResult, placeholder, queue, newQueue) match {
+          case ShuffledQueueAndLease(Some(lease), nq) => DequeueResult(Option(LeasedActor(placeholder.actor, lease)), copy(queue = nq))
+          case ShuffledQueueAndLease(None, nq) => DequeueResult(None, copy(queue = nq))
+        }
+      case None =>
+        DequeueResult(None, this)
+    }
+  }
+
+  private def shuffleQueueAndGetLease(result: UnhoggableTokenPoolResult,
```

Comment:
Can you explain this ?

Reply:
Specifically the shuffle part, why does the hog get moved at the end ?

Reply:
This is a pretty bad name - it's really "try to unpack a lease from this result, and depending on the result, choose what the new queue should look like"

Reply:
The "hogs go to the back" rule is intended to be an approximation of "you stay at the back of the queue for as long as you're still using all your available tokens"